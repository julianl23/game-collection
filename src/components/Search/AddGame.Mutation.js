import gql from "graphql-tag";

const AddGameToCollection = gql`
  mutation AddGameToCollection($input: GameInput!) {
    AddGameToCollection(input: $input) {
      _id
    }
  }
`;

export default AddGameToCollection;
