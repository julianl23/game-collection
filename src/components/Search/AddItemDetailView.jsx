import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { theme } from "styled-tools";
import { Formik } from "formik";

import Button from "../Button";
import Checkbox from "../Checkbox";
import TextArea from "../TextArea";
import TextField from "../TextField";
import { PlusSquare } from "../Icons";

// TODO: This is reused in two places
const AddButton = styled(Button)`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 12px;

  svg {
    margin-right: 5px;
  }
`;

const AddItemDetailViewContainer = styled.div`
  display: flex;
`;

const PlatformSelectLabel = styled.label`
  display: flex;
  flex-direction: column;
  margin: 0 10px 20px 0;
  flex: 1;
  width: 100%;
`;

const PlatformSelect = styled.select`
  height: 36px;
  background: ${theme("purpleWhite")};
  margin: 10px 0 0;
`;

const AddForm = styled.form`
  width: 100%;
`;

const ButtonWrapper = styled.div``;

class AddItemDetailView extends Component {
  validateForm = values => {
    const { game } = this.props;
    const errors = {};

    if (game.platforms.length > 1 && !values.platform) {
      errors.platform = "Please select a platform to add.";
    }

    if (typeof values.cost !== "number") {
      errors.cost = "Please enter a numeric value for cost.";
    }

    return errors;
  };

  handleSubmitForm = (
    values,
    { setSubmitting, setErrors /* setValues and other goodies */ }
  ) => {
    const { handleAddGame, game } = this.props;
    const { notes, cost, cartdisk, casebox, manual, other, platform } = values;

    // need to set submitting and set errors

    handleAddGame({
      notes,
      cost,
      hasCartDisk: cartdisk,
      hasCaseBox: casebox,
      hasManual: manual,
      hasOtherInserts: other,
      selectedPlatform:
        game.platforms.length === 1 ? game.platforms[0] : platform,
    });
  };

  render() {
    const { game, handleToggleAddView } = this.props;
    const { platforms } = game;

    return (
      <AddItemDetailViewContainer>
        <Formik
          initialValues={{
            notes: "",
            cost: 0,
            cartdisk: false,
            casebox: false,
            manual: false,
            other: false,
            platform: "",
          }}
          validate={this.validateForm}
          onSubmit={this.handleSubmitForm}
          render={({
            values,
            errors,
            handleChange,
            handleSubmit,
            isSubmitting,
          }) => (
            <AddForm onSubmit={handleSubmit}>
              {platforms.length > 1 && (
                <PlatformSelectLabel htmlFor="platform">
                  Platform
                  <PlatformSelect
                    id="platform"
                    name="platform"
                    onChange={handleChange}
                    value={values.platform}
                  >
                    <option>Select</option>
                    {platforms.map(platform => (
                      <option key={platform._id} value={platform._id}>
                        {platform.name}
                      </option>
                    ))}
                  </PlatformSelect>
                  {errors.platform}
                </PlatformSelectLabel>
              )}
              <div>
                <TextArea
                  id="notes"
                  name="notes"
                  value={values.notes}
                  labelText="Notes"
                  onChange={handleChange}
                />
              </div>
              <div>
                <TextField
                  type="number"
                  id="cost"
                  name="cost"
                  labelText="Cost"
                  onChange={handleChange}
                  value={values.cost}
                  errorMessage={errors.cost}
                />
              </div>
              <div>
                <Checkbox
                  id="cartdisk"
                  name="cartdisk"
                  value={values.cartdisk}
                  onChange={handleChange}
                  labelText="Has cart/disk?"
                />
                <Checkbox
                  id="casebox"
                  name="casebox"
                  value={values.casebox}
                  onChange={handleChange}
                  labelText="Has case/box?"
                />
                <Checkbox
                  id="manual"
                  name="manual"
                  value={values.manual}
                  onChange={handleChange}
                  labelText="Has manual?"
                />
                <Checkbox
                  id="other"
                  name="other"
                  value={values.other}
                  onChange={handleChange}
                  labelText="Has other inserts?"
                />
              </div>
              <ButtonWrapper>
                <AddButton type="submit" disabled={isSubmitting} block>
                  <PlusSquare /> Add to my collection
                </AddButton>
                <Button
                  type="button"
                  buttonStyle="secondary"
                  onClick={handleToggleAddView}
                  block
                >
                  Cancel
                </Button>
              </ButtonWrapper>
            </AddForm>
          )}
        />
      </AddItemDetailViewContainer>
    );
  }
}

AddItemDetailView.propTypes = {
  game: PropTypes.object.isRequired,
  handleAddGame: PropTypes.func.isRequired,
  handleToggleAddView: PropTypes.func.isRequired,
};

export default AddItemDetailView;
