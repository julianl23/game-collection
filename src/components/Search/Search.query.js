import { gql } from "apollo-boost";

const SEARCH = gql`
  query GameSearch($query: String!) {
    games(query: $query) {
      _id
      title
      developer {
        _id
        name
      }
      publisher {
        _id
        name
      }
      platforms {
        _id
        name
        logo {
          url
          width
          height
          cloudinary_id
        }
      }
      releaseDate
      description
      createdAt
      updatedAt
      igdbId
      cover {
        url
        width
        height
      }
      gameModes
      multiplayerModes {
        platform {
          _id
        }
        offlinecoop
        onlinecoop
        lancoop
        campaigncoop
        splitscreenonline
        splitscreen
        dropin
        offlinecoopmax
        onlinecoopmax
        onlinemax
        offlinemax
      }
    }
  }
`;

export default SEARCH;
