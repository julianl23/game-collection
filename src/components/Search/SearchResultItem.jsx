import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { theme } from "styled-tools";

import Button from "../Button";
import AddItemDetailView from "./AddItemDetailView";
import { PlusSquare } from "../Icons";

const Item = styled.li`
  margin-bottom: 48px;
`;

const AddButton = styled(Button)`
  display: flex;
  justify-content: center;
  align-items: center;

  svg {
    margin-right: 5px;
  }
`;

const SearchResultInfo = styled.div`
  display: flex;
  margin-bottom: 20px;
`;

const ContentSection = styled.div`
  margin-left: 13px;
`;

const Companies = styled.p`
  margin: 0;
`;

const Title = styled.span`
  font-size: 16px;
  font-weight: 500;
  line-height: 19px;
  color: ${theme("deepPurple")};
  margin-bottom: 7px;
  display: inline-block;
`;

const Platforms = styled.p`
  margin: 20px 0 0;
`;

// const CancelButton = styled.button`
//   -webkit-appearance: none;
//   background: transparent;
//   border: 0;
//   align-self: flex-start;
//   padding: 5px;
//   cursor: pointer;
// `;

class SearchResultItem extends Component {
  constructor() {
    super();
    this.state = {
      expanded: false,
    };
  }

  renderCompanyList = (companies, prefix) => {
    const companyNames = companies.map(company => company.name);
    let companyNameString;

    if (!companies.length) {
      return null;
    }

    if (companies.length > 1) {
      const otherCount = companies.length - 2;
      const othersString = `${otherCount} ${otherCount > 1 && "s"}`;
      companyNameString = `${companyNames[0]}, ${
        companyNames[1]
      }, and ${othersString}`;
    } else {
      companyNameString = companyNames.join(", ");
    }

    return (
      <React.Fragment>
        {prefix} {companyNameString}
      </React.Fragment>
    );
  };

  renderPlatformList = platforms => {
    const platformNames = platforms.map(platform => platform.name);
    let platformsString;

    if (!platforms.length) {
      return null;
    }

    if (platforms.length > 2) {
      const otherCount = platforms.length - 2;
      const othersString = `${otherCount} other${otherCount > 1 ? "s" : ""}`;
      platformsString = `${platformNames[0]}, ${
        platformNames[1]
      }, and ${othersString}`;
    } else {
      platformsString = platformNames.join(", ");
    }

    return <Platforms>Platforms: {platformsString}</Platforms>;
  };

  handleToggleAddView = () => {
    const { expanded } = this.state;
    this.setState({
      expanded: !expanded,
    });
  };

  handleAddGame = collectionItem => {
    console.log(collectionItem);
  };

  render() {
    const { game } = this.props;
    const { expanded } = this.state;

    const { cover, title, developer, publisher, platforms } = game;

    return (
      <Item>
        <SearchResultInfo>
          <img src={cover.url} alt={`Cover of ${title}`} />
          <ContentSection>
            <Title>{title}</Title>
            <Companies>
              {this.renderCompanyList(developer, "Developed by")}
              <br />
              {this.renderCompanyList(publisher, "Published by")}
            </Companies>

            {this.renderPlatformList(platforms)}
          </ContentSection>
        </SearchResultInfo>
        {expanded && (
          <AddItemDetailView
            game={game}
            handleAddGame={this.handleAddGame}
            handleToggleAddView={this.handleToggleAddView}
          />
        )}
        {!expanded && (
          <AddButton block onClick={this.handleToggleAddView}>
            <PlusSquare />Add to my collection
          </AddButton>
        )}
      </Item>
    );
  }
}

SearchResultItem.propTypes = {
  game: PropTypes.object.isRequired,
};

export default SearchResultItem;
