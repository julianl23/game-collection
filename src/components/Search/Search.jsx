import React, { Component } from "react";
import { ApolloConsumer } from "react-apollo";
import styled from "styled-components";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";

import TextField from "../TextField";
import Button from "../Button";
import SearchResultItem from "./SearchResultItem";
import SEARCH from "./Search.query";

const SearchSection = styled.section`
  grid-column: 1 / span 4;
`;

const SearchResultsList = styled.ul`
  padding: 0;
  list-style-type: none;
`;

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      query: "",
      results: null,
    };
  }

  componentDidMount() {
    const { match } = this.props;
    if (match.params && match.params.query) {
      this.setState({
        query: match.params.query,
        searchOnFirstOpen: true,
      });
    }
  }

  handleChange = e => {
    this.setState({
      query: e.target.value,
    });
  };

  handleSearchSubmit = async (e, client) => {
    if (e) e.preventDefault();

    this.setState({ searchOnFirstOpen: false });

    const { query } = this.state;

    try {
      const { data } = await client.query({
        query: SEARCH,
        variables: { query },
      });

      this.setState({
        results: data.games,
      });
    } catch (err) {
      console.log(err);
    }
  };

  render() {
    const { query, results, searchOnFirstOpen } = this.state;

    return (
      <ApolloConsumer>
        {client => {
          if (searchOnFirstOpen) {
            this.handleSearchSubmit(null, client);
          }

          return (
            <SearchSection>
              <h1>Search</h1>
              <form
                action="/search"
                method="POST"
                onSubmit={e => this.handleSearchSubmit(e, client)}
              >
                <TextField
                  id="query"
                  value={query}
                  labelText=""
                  onChange={this.handleChange}
                />
                <Button type="submit">Search</Button>
              </form>
              {results && (
                <SearchResultsList>
                  {results.map(game => (
                    <SearchResultItem game={game} key={game._id} />
                  ))}
                </SearchResultsList>
              )}
            </SearchSection>
          );
        }}
      </ApolloConsumer>
    );
  }
}

Search.propTypes = {
  match: PropTypes.object.isRequired,
};

export default withRouter(Search);
