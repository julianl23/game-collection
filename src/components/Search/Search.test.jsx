import React from "react";
import TestRenderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { MockedProvider } from "react-apollo/test-utils";

import Search from "./Search";

describe("Search", () => {
  it("renders correctly", () => {
    const instance = TestRenderer.create(
      <BrowserRouter>
        <MockedProvider>
          <Search />
        </MockedProvider>
      </BrowserRouter>
    );
    expect(instance.toJSON()).toMatchSnapshot();
  });

  it("needs to have test cases added", () => {
    expect(true).toBe(true);
  });
});
