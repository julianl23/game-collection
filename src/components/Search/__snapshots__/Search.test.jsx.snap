// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
.c1 {
  margin-bottom: 20px;
}

.c2 {
  display: block;
  margin-bottom: 10px;
}

.c3 {
  display: block;
  border: 1px solid;
  padding: 15px 10px;
  height: 45px;
  width: 100%;
}

.c4 {
  height: 36px;
  background: #322F6F;
  border-radius: 3px;
  font-size: 14px;
  font-weight: 300;
  line-height: 16px;
  color: #fff;
  cursor: pointer;
  display: inline-block;
}

.c0 {
  grid-column: 1 / span 4;
}

<section
  className="c0"
>
  <h1>
    Search
  </h1>
  <form
    action="/search"
    method="POST"
    onSubmit={[Function]}
  >
    <div
      className="c1"
    >
      <label
        className="c2"
        htmlFor="query"
      >
        Search
      </label>
      <div
        className="textfield-wrapper"
      >
        <input
          className="c3"
          id="query"
          name=""
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      
    </div>
    <button
      className="c4"
      onClick={null}
      size="normal"
      type="submit"
    >
      Search
    </button>
  </form>
</section>
`;
